asaplib/cluster/ml_cluster_fit.py

```diff
@@ -100,115 +100,267 @@ def pack(self):
+        -------
         """
-        plt.scatter(rho, delta)
-        plt.plot([min(rho), max(rho)], [self.deltamin, self.deltamin], c='red')
-        plt.plot([self.rhomin, self.rhomin], [min(delta), max(delta)], c='red')
-        plt.xlabel('rho')
-        plt.ylabel('delta')
-        plt.show()
-        ###
-
-        nclust = 0
-        cl = np.zeros(len(rho), dtype='int') - 1
-        for i in range(len(rho)):
-            # we select cluster centers {cl} with rho({c}) > rhomin && delta({c}) > delta_min
-            if rho[i] > self.rhomin and delta[i] > self.deltamin:
-                nclust += 1
-                cl[i] = nclust
-
-        ### Assignment
-        ordrho = np.argsort(rho)[::-1]
-        rho_ord = rho[ordrho]
-        for i in range(len(rho)):
-            if cl[ordrho[i]] == -1:
-                cl[ordrho[i]] = cl[nneigh[ordrho[i]]]
-        return cl
-
-    def estimate_delta(self, dist, rho):
-
-        ''' For each data point i, compute the distance (delta_i) between i and j,
-           j is the closet data point that has a density higher then i, i.e. rho(j) > rho(i).
-        '''
-
-        delta = (rho * 0.0).copy()
-        nneigh = np.ones(len(delta), dtype='int')
-        for i in range(len(rho)):
-            # for data i, find all points that have higher density
-            js = np.where(rho > rho[i])[0]
-            if len(js) == 0:
-                # if there's no j's that have higher density than i, we set delta_i to be a large distance
-                delta[i] = np.max(dist[i, :])
-                nneigh[i] = i
+        ordered = np.argsort(-self.dens)  # in descending order of density
+        self.cluster = np.zeros(data.shape[0], dtype='int')
+        tt = np.arange(data.shape[0])
+        center_label = np.zeros(data.shape[0], dtype='int')
+        ncluster = -1
+        for i in range(data.shape[0]):
+            j = ordered[i]
+            if (self.dens[j] > self.dens_cut) & (self.delta[j] > self.delta_cut):
+                ncluster = ncluster + 1
+                self.cluster[j] = ncluster
+                center_label[j] = ncluster
             else:
-                # find the nearest j that has higher density then i
-                delta[i] = np.min(dist[i, js])
-                nneigh[i] = js[np.argmin(dist[i, js])]
-        return delta, nneigh
+                self.cluster[j] = self.cluster[self.ref[j]]
+                center_label[j] = -1
+        self.centers = tt[(center_label != -1)]
+        bord = np.zeros(data.shape[0], dtype='int')
+        self.halo = np.copy(self.cluster)
+
+        for i in range(data.shape[0]):
+            for j in self.indices[i, :][(self.distances[i, :] <= self.dc)]:
+                if self.cluster[i] != self.cluster[j]:
+                    bord[i] = 1
+        halo_cutoff = np.zeros(ncluster + 1)
+        for i in range(ncluster + 1):
+            dd = self.dens[((bord == 1) & (self.cluster == i))]
+            halo_cutoff[i] = np.max(dd)
```

Comment:
This is the line that's currently failing with the Ni example. @BingqingCheng would be useful if you could reproduce the error.